worker/src/handlers.rs

```diff
@@ -73,4 +240,140 @@ impl PrimaryToWorker for PrimaryReceiverHandler {
 
         Ok(anemo::Response::new(()))
     }
+    async fn synchronize(
+        &self,
+        request: anemo::Request<WorkerSynchronizeMessage>,
+    ) -> Result<anemo::Response<()>, anemo::rpc::Status> {
+        let message = request.into_body();
+
+        let mut missing = HashSet::new();
+        let mut available = HashSet::new();
+
+        for digest in message.digests.iter() {
+            // Check if we already have the batch.
+            match self.store.read(*digest).await {
+                Ok(None) => {
+                    missing.insert(*digest);
+                    debug!("Requesting sync for batch {digest}");
+                }
+                Ok(Some(_)) => {
+                    available.insert(*digest);
+                    trace!("Digest {digest} already in store, nothing to sync");
+                    continue;
+                }
+                Err(e) => {
+                    error!("{e}");
+                    continue;
+                }
+            };
+        }
+
+        // reply back immediately for the available ones
+        if !available.is_empty() {
```

Comment:
Do we need to check this condition? the `for` loop will just do nothing if available is empty right?

Reply:
Good point, this is code moved without modification from synchronizer but maybe the original reason for this check no longer applies